### DISCUSSION TITLE
Asynchronous event listeners

### DISCUSSION THREAD

**Discussion by Kodiologist** [pyodide/pyodide#2229] [Q&A] [answered]
When I use an asynchronous function as an event listener, it doesn't seem to get run. I'm not sure if there's a bug or something about asynchronous semantics in JavaScript that I don't understand.

For example, if in [the console](https://brython.info/console.html) I enter

```python
import js, pyodide
def f(event): js.console.log('clicked')
js.document.addEventListener('click', pyodide.create_proxy(f))
```

and then click the page, I can see "clicked" in the browser console. On the other hand, if I change the `def` to `async def`, clicking the page appears to do nothing, and sometimes I see the warning `sys:1: RuntimeWarning: coroutine 'f' was never awaited`.

**Comment by hoodmane** [comment, marked as answer]
I think this is confusing enough people that it should probably be considered a bug in Pyodide, so hopefully we'll have a fix in 0.20. If you want your code to work, you need to say:
```py
def wrapper(e):
   asyncio.ensure_future(async_method(e))
```
And then use wrapper instead. If you're doing this a lot you'll likely want to make a decorator.

**Reply by Kodiologist** [reply]
Thanks. Whence can I import `async_method`? Is that example missing a `return`? And would I use it like `js.document.addEventListener('click', pyodide.create_proxy(wrapper(f)))`?

**Reply by hoodmane** [reply]
Well it should probably more accurately be:
```py
from functools import wraps
import asyncio
def wrapper(f):
	@wraps(f)
	def inner(*args, **kwargs):
		return asyncio.ensure_future(f(*args, **kwargs))
	return inner
```
and then you could use it like 
```py
async def f(e):
   # ... do stuff with e
js.document.addEventListener('click', pyodide.create_proxy(wrapper(f)))
```
or like 
```py
@wrapper
async def f(e):
   # ... do stuff with e
js.document.addEventListener('click', pyodide.create_proxy(f))
```
